### PR TITLE
[pruning][core][feature] Implement convert for pruner

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97545

Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`

### CODE REVIEW

File: torch/ao/pruning/sparsifier/utils.py [hunk 1]
```diff
@@ -1,16 +1,70 @@
...
+
+        # respect device affinity when swapping modules
+        devices = {p.device for p in chain(mod.parameters(), mod.buffers())}
+        assert len(devices) <= 1, (
+            "swap_module only works with cpu or single-device CUDA modules, "
+            "but got devices {}".format(devices)
+        )
+        device = next(iter(devices)) if len(devices) > 0 else None
+        if device:
+            new_mod.to(device)
```
Comment (jerryzh168) [inline torch/ao/pruning/sparsifier/utils.py:60]: nit: looks like this is copied from eager mode quantization? maybe you can refactor this to a separate helper function?

File: torch/ao/pruning/sparsifier/utils.py [hunk 2]
```diff
@@ -1,16 +1,70 @@
...
+    "swap_module",
     "module_to_fqn",
     "fqn_to_module",
     "get_arg_info_from_tensor_fqn",
     "FakeSparsity",
 ]
 
 
-def module_to_fqn(model: nn.Module, module: nn.Module, prefix: str = "") -> Optional[str]:
+def module_contains_param(module, parametrization):
```
Comment (jerryzh168) [inline torch/ao/pruning/sparsifier/utils.py:17]: nit: typing

File: torch/ao/pruning/sparsifier/utils.py [hunk 3]
```diff
@@ -1,16 +1,70 @@
...
+        return any(
+            [
+                any(isinstance(param, parametrization) for param in param_list)
+                for key, param_list in module.parametrizations.items()
+            ]
+        )
+    return False
+
+
+def swap_module(mod, mapping):
```
Comment (jerryzh168) [inline torch/ao/pruning/sparsifier/utils.py:29]: nit: typing

File: torch/ao/pruning/sparsifier/base_sparsifier.py [hunk 4]
```diff
@@ -288,12 +298,45 @@ def squash_mask(self,
                 # TODO handle multiple tensor being quantized on a single module, where to store sparse_params?
                 module.sparse_params = sparse_params
 
-    def convert(self):
-        # TODO: Call the torch.ao.utils.convert in here
-        raise NotImplementedError(
-            "`convert` is not implemented. Please, use "
-            "`torch.ao.utils.convert` instead."
-        )
+    def convert(
```
Comment (jerryzh168) [inline torch/ao/pruning/sparsifier/base_sparsifier.py:301]: please add types as well

**Review by jerryzh168 — APPROVED** [review]
LGTM

File: torch/ao/pruning/sparsifier/utils.py (hunk 1 above)
Comment (jcaip) [reply torch/ao/pruning/sparsifier/utils.py:60]: I'll add this to the BE tracker and take another look later. 

I'm not sure atm what the best way to reuse the shared code other than passing a function around for the sparse/quant specific steps. And I don't like that b/c I feel like it makes the code harder to read.

Not sure if you have any opinions on this?

**Pull request closed without merging**